File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0412_Fizz_Buzz.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0412_Fizz_Buzz {
    public List<String> fizzBuzz(int n) {
        if (n <= 0) {
            return null;
        }
        int three = 3, five = 5;
        List<String> res = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            --three;
            --five;
            if (three == 0 && five == 0) {
                res.add("FizzBuzz");
                three = 3;
                five = 5;
            } else if (three == 0) {
                res.add("Fizz");
                three = 3;
            } else if (five == 0) {
                res.add("Buzz");
                five = 5;
            } else {
                res.add(i + "");
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0412_Fizz_Buzz.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<string> fizzBuzz(int n) {
    if (n <= 0) {
      return vector<string>();
    }
    vector<string> res;
    int three = 3, five = 5;
    for (int i = 1; i <= n; i++) {
      three--; five--;
      if (three == 0 && five == 0) {
        res.push_back("FizzBuzz");
        three = 3;
        five = 5;
      } else if (three == 0) {
        res.push_back("Fizz");
        three = 3;
      } else if (five == 0) {
        res.push_back("Buzz");
        five = 5;
      } else {
        res.push_back(to_string(i));
      }
    }
    return res;
  }
};
